### ISSUE TITLE
Compiler breaks when constant name is the same as type name

### ISSUE THREAD

**Issue by deimastep** [dotnet/roslyn#19458]
**Version Used**: VS 2017 version 15.2 release 26430.4

**Steps to Reproduce**:

1. Start build this code in .Net Core console project "Test"
```csharp
using System.Text.RegularExpressions;
// ...
const RegexOptions RegexOptions = RegexOptions.CultureInvariant | RegexOptions.IgnoreCase;
var regex = new Regex(@"\s*", RegexOptions);
```

**Expected Behavior**:
Should be compiled.
If I remove `const` everything is fine:
```csharp
RegexOptions RegexOptions = RegexOptions.CultureInvariant | RegexOptions.IgnoreCase;
var regex = new Regex(@"\s*", RegexOptions);
```

**Actual Behavior**:
Compiler just emits message:
`CSC : error CS7038: Failed to emit module 'Test'.`

**Comment by jcouv** [comment]
@deimastep I was not able to reproduce this issue.
I created a .NET Core project in Visual Studio 2015 version 15.3 (preview) with the code below. The IDE and build behave correctly (no errors).
```C#
using System.Text.RegularExpressions;

public class C
{
    void M()
    {
        const RegexOptions RegexOptions = RegexOptions.CultureInvariant | RegexOptions.IgnoreCase;
        var regex = new Regex(@"\s*", RegexOptions);
    }
}
```

I also tried making the const a field, then I got an error (`error CS0110: The evaluation of the constant value for 'C.RegexOptions' involves a circular definition`):

```C#
using System.Text.RegularExpressions;

public class C
{
    // error CS0110: The evaluation of the constant value for 'C.RegexOptions' involves a circular definition
    const RegexOptions RegexOptions = RegexOptions.CultureInvariant | RegexOptions.IgnoreCase;
    void M()
    {
        var regex = new Regex(@"\s*", RegexOptions);
    }
}
```

But I was not able to get the "failed to emit module" error you ran into.

**Comment by deimastep** [comment]
@jcouv  I believe you mean Visual Studio **2017** version 15.3 (preview), not a Visual Studio **2015**?

Unfortunately I didn't tried 15.3 preview. But on version 15.2 your example with local constant fails to compile in the same way.
So I'm making a conclusion that this bug is fixed in 15.3 version.

**Comment by svick** [comment]
I can reproduce this on VS 2017, VS 15.3 Preview and Roslyn `master`, so I don't think it's been fixed.

I believe one issue is [this line](https://github.com/dotnet/roslyn/blob/166ff84/src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs#L927) which converts warnings in `const` assignments to errors.

@jcouv Are you sure you tried it in a .Net Core application?

**Comment by jcouv** [comment]
Thanks @svick for the fix! #20729
